---
title: "NovClusterprofiler184spl"
author: "Saul"
date: "2023-11-16"
output: html_document
---

```{r}
#save(DvB_uniqgenes, file = "DvB_uniqgenes.RData")
# Change directory
# setwd("C:/Users/SaulKarr/OneDrive - Instituto Tecnologico y de Estudios Superiores de Monterrey/DBT_data/neubulkRNA")

#https://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html

#Usaremos el df con los geneIDs de ENSEMBLE en rownames

#BiocManager::install("clusterProfiler", version = 3.18)
#BiocManager::install("gprofiler2")
#BiocManager::install("pathview")
#BiocManager::install("enrichplot")
#install.packages("devtools")
#install.packages("cowplot")
#install.packages("patchwork")
library(clusterProfiler)
library(gprofiler2)
library(enrichplot)
# we use ggplot2 to add x axis labels (ex: ridgeplot)
library(ggplot2)
library(cowplot)
library(fgsea)
library(patchwork)
library(stats)

# SET THE DESIRED ORGANISM
#https://bioconductor.org/packages/release/BiocViews.html#___OrgDb

organism = "org.Hs.eg.db"

#BiocManager::install(organism, character.only = TRUE)
library(organism, character.only = TRUE)

#Prepare Input
indir <- "C:/Users/ecoss/OneDrive - CINVESTAV/Documentos/Posdoc_LIIGH/Ayuda_extra/Saul_Karr_neutrofilos/Analysis_graph_saul/data/"
# data from deseq2
load(paste0(indir, "DBvB_uniqprotcodgenes.RData")) # DvB_uniqgenes
load(paste0(indir,"SexFvM_uniqprotcodgenes.RData")) # SexFvM_uniqgenes
load(paste0(indir,"AgeOvY_uniqprotcodgenes.RData")) # AgeOvY_uniqgenes
load(paste0(indir,"Cluster1_uniqprotcodgenes.RData")) # Cluster1_uniqgenes
load(paste0(indir,"Cluster2_uniqprotcodgenes.RData")) # Cluster2_uniqgenes
load(paste0(indir,"Cluster3_uniqprotcodgenes.RData")) # Cluster3_uniqgenes

# Eliminar esta linea de codigo
# DvB_uniqgenes <- DBvB_uniqgenes

# we want the log2 fold change 
original_gene_listDvB    <- DvB_uniqgenes$log2FoldChange
original_gene_listSexFvM <- SexFvM_uniqgenes$log2FoldChange
original_gene_listAgeOvY <- AgeOvY_uniqgenes$log2FoldChange
original_gene_listC1 <- Cluster1_uniqgenes$log2FoldChange
original_gene_listC2 <- Cluster2_uniqgenes$log2FoldChange
original_gene_listC3 <- Cluster3_uniqgenes$log2FoldChange

# name the vector
names(original_gene_listDvB) <- DvB_uniqgenes$external_gene_name
names(original_gene_listSexFvM) <- SexFvM_uniqgenes$external_gene_name
names(original_gene_listAgeOvY) <- AgeOvY_uniqgenes$external_gene_name
names(original_gene_listC1) <- Cluster1_uniqgenes$external_gene_name
names(original_gene_listC2) <- Cluster2_uniqgenes$external_gene_name
names(original_gene_listC3) <- Cluster3_uniqgenes$external_gene_name

# omit any NA values 
gene_listDvB<-na.omit(original_gene_listDvB)
gene_listSexFvM<-na.omit(original_gene_listSexFvM)
gene_listAgeOvY<-na.omit(original_gene_listAgeOvY)
gene_listC1<-na.omit(original_gene_listC1)
gene_listC2<-na.omit(original_gene_listC2)
gene_listC3<-na.omit(original_gene_listC3)

# sort the list in decreasing order (required for clusterProfiler)
gene_listDvB = sort(gene_listDvB, decreasing = TRUE)
gene_listSexFvM = sort(gene_listSexFvM, decreasing = TRUE)
gene_listAgeOvY = sort(gene_listAgeOvY, decreasing = TRUE)
gene_listC1 = sort(gene_listC1, decreasing = TRUE)
gene_listC2 = sort(gene_listC2, decreasing = TRUE)
gene_listC3 = sort(gene_listC3, decreasing = TRUE)

## Gene Set Enrichment
# Params:  
  
#**keyType** This is the source of the annotation (gene ids). The options vary for each annotation. In the example of  *org.Dm.eg.db*, the options are:   
  
#"ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS" "ENTREZID"      
#"ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "FLYBASE"      "FLYBASECG"    "FLYBASEPROT"   
#"GENENAME"     "GO"           "GOALL"        "MAP"          "ONTOLOGY"     "ONTOLOGYALL"   
#"PATH"         "PMID"         "REFSEQ"       "SYMBOL"       "UNIGENE"      "UNIPROT"  
  
# Check which options are available with the `keytypes` command, for example `keytypes(org.Hs.eg.db)`. 

keytypes(org.Hs.eg.db)
  
#**ont** one of "BP", "MF", "CC" or "ALL"  
#**nPerm** permutation numbers, the higher the number of permutations you set, the more accurate your results is, but it will also cost longer time for running permutation.  
#**minGSSize** minimal size of each geneSet for analyzing.   
#**maxGSSize** maximal size of genes annotated for testing.   
#**pvalueCutoff** pvalue Cutoff.   
#**pAdjustMethod** one of "holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr", "none" 

gseDvB <-    gseGO(gene_listDvB, ont = "BP", keyType = "SYMBOL", OrgDb = "org.Hs.eg.db",
                   minGSSize    = 10, maxGSSize    = 500, # min genes, max genes. pvalue
                   pvalueCutoff = 0.05, verbose = FALSE)
gseSexFvM <- gseGO(gene_listSexFvM, ont = "BP", keyType = "SYMBOL", OrgDb = "org.Hs.eg.db",
                   minGSSize    = 10, maxGSSize    = 500, # min genes, max genes. pvalue
                   pvalueCutoff = 0.05, verbose = FALSE)
gseAgeOvY <- gseGO(gene_listAgeOvY, ont = "BP", keyType = "SYMBOL", OrgDb = "org.Hs.eg.db",
                   minGSSize    = 10, maxGSSize    = 500, # min genes, max genes. pvalue
                   pvalueCutoff = 0.05, verbose = FALSE)
gseC1 <-     gseGO(gene_listC1, ont = "BP", keyType = "SYMBOL", OrgDb = "org.Hs.eg.db",
                   minGSSize    = 10, maxGSSize    = 500, # min genes, max genes. pvalue
                   pvalueCutoff = 0.05, verbose = FALSE)
gseC2 <-     gseGO(gene_listC2, ont = "BP", keyType = "SYMBOL", OrgDb = "org.Hs.eg.db",
                   minGSSize    = 10, maxGSSize    = 500, # min genes, max genes. pvalue
                   pvalueCutoff = 0.05, verbose = FALSE)
gseC3 <-     gseGO(gene_listC3, ont = "BP", keyType = "SYMBOL", OrgDb = "org.Hs.eg.db",
                   minGSSize    = 10, maxGSSize    = 500, # min genes, max genes. pvalue
                   pvalueCutoff = 0.05, verbose = FALSE)

# Save info
outdir <- "C:/Users/ecoss/OneDrive - CINVESTAV/Documentos/Posdoc_LIIGH/Ayuda_extra/Saul_Karr_neutrofilos/Analysis_graph_saul/outputs/"

save(gseDvB, file =  paste0(outdir, "gseDvB.RData"))
save(gseSexFvM, file = paste0(outdir, "gseSexFvM.RData"))
save(gseAgeOvY, file =paste0(outdir, "gseAgeOvY.RData"))
save(gseC1, file = paste0(outdir, "gseC1.RData"))
save(gseC2, file = paste0(outdir, "gseC2.RData"))
save(gseC3, file = paste0(outdir, "gseC3.RData"))

# remove data
rm(list = ls())
```

## Dotplot

```{r gseDvB data, echo=TRUE, fig.width=15, fig.height=8}
# Load data
outdir <- "C:/Users/ecoss/OneDrive - CINVESTAV/Documentos/Posdoc_LIIGH/Ayuda_extra/Saul_Karr_neutrofilos/Analysis_graph_saul/outputs/"

load(file = paste0(outdir, "gseDvB.RData")) # gseDvB
load(file = paste0(outdir, "gseSexFvM.RData"))
load(file = paste0(outdir, "gseAgeOvY.RData"))
load(file = paste0(outdir,  "gseC1.RData"))
load(file = paste0(outdir,  "gseC2.RData"))
load(file = paste0(outdir,  "gseC3.RData"))
```


```{r Grafica de Saul,echo=TRUE, fig.width=15, fig.height=8}
# Solo muestra los 10 primeros
# png("dp1.png", res = 300, width = 15, height = 20, units = "cm")

dp1 <- dotplot(gseDvB, showCategory=10, split=".sign", # dividir en up and downregulated
               font.size = 9, title = "") + facet_grid(.~.sign)+
  theme(axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))+ theme(legend.position="right", legend.text = element_text(size = 8),legend.title = element_text(size = 8)) + labs(title = "GO terms", y = "", x="") + theme(plot.title = element_text(size=10, hjust=0.5)) #+ scale_size_continuous(limits = c(0, 100), breaks = c(0, 25, 50, 75, 100))

dp1

# dev.off()
```

```{r processGOTerms function}
library(clusterProfiler)
library(ggplot2)
library(rrvgo) #revigo
library(DOSE)
library(tidyverse)
library(org.Hs.eg.db)

processGOTerms <- function(gseDvB, orgdb="org.Hs.eg.db", ont="BP", method="Rel", threshold=0.7, 
                           output_dir = ".", file_name = "dp1", 
                           width = 10, height = 8, dpi = 300) {
  
  # Convertir a data.frame
  gse_df <- as.data.frame(gseDvB)
  
  # Añadir información sobre la expresión
  message("Adding information about gene expression")
  gse_df <- gse_df %>%
    dplyr::select(ID, Description, enrichmentScore, p.adjust, qvalue, core_enrichment) %>%
    dplyr::mutate(Significance = ifelse(enrichmentScore > 0, "Upregulated", "Downregulated"))
  
  # Función para contar el número de genes
  message("Counting number of genes")
  count_genes <- function(text) {
    words <- unlist(strsplit(text, split = "/"))
    return(length(words))
  }
  
  # Añadir información sobre el número de genes
  gse_df <- gse_df %>%
    mutate(genes_number = sapply(core_enrichment, count_genes))
  
  # Calcular la matriz de similitud
  message("Reduce GO terms with Revigo")
  simMatrix <- calculateSimMatrix(gse_df$ID,
                                  orgdb=orgdb,
                                  ont=ont,
                                  method=method)
  
  # Calcular las puntuaciones
  scores <- setNames(-log10(gse_df$qvalue), gse_df$ID)
  
  # Reducir los términos de GO
  reducedTerms <- reduceSimMatrix(simMatrix,
                                  scores,
                                  threshold=threshold,
                                  orgdb=orgdb)
  
  # Crear una lista de términos y sus respectivas familias
  term_family_list <- reducedTerms %>%
    dplyr::select(go, term, parentTerm) %>%
    distinct()
  
  # Renombrar la columna GOterm a ID para unirse fácilmente
  term_family_list <- term_family_list %>%
    rename(ID = go)
  
  # Realizar la unión y llenar los términos faltantes
  gseDvBcombined_data <- gse_df %>%
    left_join(term_family_list, by = "ID") %>%
    drop_na(parentTerm) %>% 
    mutate(parentTerm = if_else(is.na(parentTerm), term, parentTerm))
  
  # Generar y guardar la figura
  message("Graph")
  
  # Cambiar de dotplot a barplot
  
  p <- ggplot(gseDvBcombined_data, aes(x = reorder(parentTerm, -genes_number), y = enrichmentScore, fill = Significance)) +
    geom_bar(stat = "identity", position = "identity") +
    coord_flip() +
    scale_fill_manual(values = c("Upregulated" = "red", "Downregulated" = "blue")) +
    theme_minimal(base_size = 11) +
    theme(
      plot.background = element_rect(fill = "white", color = NA), 
      panel.background = element_rect(fill = "white", color = NA),
      axis.title.y = element_blank(),
      panel.spacing = unit(2, "lines"),
      strip.text = element_text(size = 12),
      axis.text.y = element_text(size = 10),
      axis.text.x = element_text(size = 10)
    ) +
    labs(
      y = "Enrichment Score",
      fill = "Significance"
    )
  
  # Definir el nombre del archivo de salida
  output_file <- file.path(output_dir, paste0("GO_Enrichment_", file_name, "_", Sys.Date(), ".png"))
  
  # Guardar la figura
  ggsave(output_file, plot = p, width = width, height = height, dpi = dpi)
  
  # Regresar la salida
  message("Return file")
  return(gseDvBcombined_data)
}

# Ejemplo de uso con un dataset
# resultado <- processGOTerms(gseDvB, output_dir = "ruta/donde/guardar", width = 12, height = 10, dpi = 400)
```


```{r Prueba}
gseDvB_list <- processGOTerms(gseDvB, output_dir = outdir, file_name = "gseDvB_BP", width = 12, height = 10, dpi = 400)
gseDvB_list

# Verificar lo que obtuvimos
# antes habian:
unique(gse_df$Description) %>% length() #1027
# Ahora hay: 
unique(gseDvB_list$parentTerm) %>% length() # 103 terminos acomodados en familias
```


## Correr el codigo para todas las comparaciones

```{r Goterms de todas las comparaciones}
# Lista de los objetos
gse_list <- list(gseDvB, gseSexFvM, gseAgeOvY, gseC1, gseC2, gseC3)

# Nombres de los archivos de salida
file_names <- c("gseDvB", "gseSexFvM", "gseAgeOvY", "gseC1", "gseC2", "gseC3")

# Inicializar una lista vacía para guardar los resultados
results_list <- list()

 # Ejecutar la función y guardar la salida en la lista
  for (i in 1:length(gse_list)) {
  message(paste("Processing file", i, "of", length(gse_list)))
  message(paste("Before:", unique(gse_list[[i]]$Description) %>% length()))
  results_list[[file_names[i]]] <- processGOTerms(gseDvB = gse_list[[i]], 
                 orgdb = "org.Hs.eg.db", 
                 ont = "BP", 
                 method = "Rel", 
                 threshold = 0.7, 
                 output_dir = outdir, 
                 file_name = file_names[i],
                 width = 10, height = 8, dpi = 300)
  message(paste("Now:",unique(gseDvB_list$parentTerm) %>% length())) #total of families
}

# Ahora, cada elemento de results_list tiene un nombre basado en file_names
# Por ejemplo, puedes acceder a la salida de "gseDvB" usando results_list[["gseDvB"]]
```
```{r}
# Ejemplo
results_list[["gseDvB"]]
```






# Aqui me quede

```{r echo=TRUE, fig.width=15, fig.height=8}
png("dp2.png", 
    res = 300, width = 15, height = 20, units = "cm")

dp2 <- dotplot(gseSexFvM, showCategory=10, split=".sign", font.size = 9, title = "") + facet_grid(.~.sign)+
  theme(axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))+
theme(legend.position="right", legend.text = element_text(size = 8),legend.title = element_text(size = 8)) + labs(title = "GO terms", y = "", x="") + theme(plot.title = element_text(size=10, hjust=0.5)) #+ scale_size_continuous(limits = c(0, 100), breaks = c(0, 25, 50, 75, 100))

dp2

dev.off()

png("dp3.png", 
    res = 300, width = 15, height = 20, units = "cm")

dp3 <- dotplot(gseAgeOvY, showCategory=10, split=".sign",  font.size = 9, title = "") + facet_grid(.~.sign)+
  theme(axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))+
theme(legend.position="right", legend.text = element_text(size = 8),legend.title = element_text(size = 8)) + labs(title = "GO terms", y = "", x="") + theme(plot.title = element_text(size=10, hjust=0.5)) #+ scale_size_continuous(limits = c(0, 100), breaks = c(0, 25, 50, 75, 100))

dp3

dev.off()

png("dp4.png", 
    res = 300, width = 15, height = 20, units = "cm")

dp4 <- dotplot(gseC1, showCategory=10, split=".sign", font.size = 9, title = "") + facet_grid(.~.sign)+
  theme(axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))+
theme(legend.position="right", legend.text = element_text(size = 8),legend.title = element_text(size = 8)) + labs(title = "GO terms", y = "", x="") + theme(plot.title = element_text(size=10, hjust=0.5)) #+ scale_size_continuous(limits = c(0, 100), breaks = c(0, 25, 50, 75, 100))

dp4

dev.off()

png("dp5.png", 
    res = 300, width = 15, height = 20, units = "cm")

dp5 <- dotplot(gseC2, showCategory=10, split=".sign", font.size = 9, title = "") + facet_grid(.~.sign)+
  theme(axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))+
theme(legend.position="right", legend.text = element_text(size = 8),legend.title = element_text(size = 8)) + labs(title = "GO terms", y = "", x="") + theme(plot.title = element_text(size=10, hjust=0.5)) #+ scale_size_continuous(limits = c(0, 100), breaks = c(0, 25, 50, 75, 100))

dp5

dev.off()

png("dp6.png", 
    res = 300, width = 15, height = 20, units = "cm")

dp6 <- dotplot(gseC3, showCategory=10, split=".sign", font.size = 9, title = "") + facet_grid(.~.sign)+
  theme(axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))+
theme(legend.position="right", legend.text = element_text(size = 8),legend.title = element_text(size = 8)) + labs(title = "GO terms", y = "", x="") + theme(plot.title = element_text(size=10, hjust=0.5)) #+ scale_size_continuous(limits = c(0, 100), breaks = c(0, 25, 50, 75, 100))

dp6

dev.off()

```

##Encrichment plot map: Enrichment map organizes enriched terms into a network with edges connecting overlapping gene sets. In this way, mutually overlapping gene sets are tend to cluster together, making it easy to identify functional modules.

```{r echo=TRUE}
#library(GOSemSim)
emapplot(gse, showCategory = 1)

#pairwise_termsim(x, method = "JC", semData = NULL, showCategory = 200)

#gene <- names(gene_list)[abs(gene_list) > 2]
#ego <- enrichGO(gene = gene,
#universe = names(gene_list),
#OrgDb = org.Hs.eg.db,
#ont = "BP",
#pAdjustMethod = "BH",
#pvalueCutoff = 0.01,
#qvalueCutoff = 0.05,
#readable = TRUE)
#d <- godata('org.Hs.eg.db', ont="BP")
#ego2 <- pairwise_termsim(ego, method="Wang", semData = d)
#emapplot(ego2)
#emapplot_cluster(ego2)

```

##Category Netplot The cnetplot depicts the linkages of genes and biological concepts (e.g. GO terms or KEGG pathways) as a network (helpful to see which genes are involved in enriched pathways and genes that may belong to multiple annotation categories).

```{r fig.width=18}

#install.packages("ggnewscale",dependencies = TRUE)
# categorySize can be either 'pvalue' or 'geneNum'
png("cnet1.png", 
    res = 300, width = 24, height = 24, units = "cm")

cnet1 <- cnetplot(gseDvB, categorySize="pvalue", foldChange=gene_listDvB, 
         showCategory = 5,
        cex_category = 0.5,
  cex_gene = 0.5,
  node_label_size = NULL,
  cex_label_category = 1,
  cex_label_gene = 1,
         circular = TRUE, colorEdge = TRUE
          )+
theme(legend.position="bottom", legend.direction = "vertical")

cnet1

dev.off()

png("cnet2.png", 
    res = 300, width = 30, height = 30, units = "cm")

cnet2 <- cnetplot(gseSexFvM, categorySize="pvalue", foldChange=gene_listSexFvM, showCategory = 5, 
         cex_category = 0.5,
  cex_gene = 0.5,
  node_label_size = NULL,
  cex_label_category = 1,
  cex_label_gene = 1,
         circular = TRUE, colorEdge = TRUE
          )+
theme(legend.position="bottom", legend.direction = "vertical")

cnet2

dev.off()

png("cnet3.png", 
    res = 300, width = 30, height = 30, units = "cm")

cnet3 <- cnetplot(gseAgeOvY, categorySize="pvalue", foldChange=gene_listAgeOvY, showCategory = 5,
        cex_category = 0.5,
  cex_gene = 0.5,
  node_label_size = NULL,
  cex_label_category = 1,
  cex_label_gene = 1,
         circular = TRUE, colorEdge = TRUE
          )+
theme(legend.position="bottom", legend.direction = "vertical")

cnet3

dev.off()

```

## Ridgeplot

Helpful to interpret up/down-regulated pathways.

```{r fig.width=18, fig.height=12}
install.packages("ggridges", dependencies = T)

ridgeplot(gse) + labs(x = "enrichment distribution")
ridgeplot(gse) + labs(x = "enrichment distribution") + theme(text = element_text(size=1)) + theme_minimal()

```

## GSEA Plot

Traditional method for visualizing GSEA result.

Params:\
**Gene Set** Integer. Corresponds to gene set in the gse object. The first gene set is 1, second gene set is 2, etc.

```{r fig.height=6}
# Use the `Gene Set` param for the index in the title, and as the value for geneSetId
gseaplot(gse, by = "all", title = gse$Description[1], geneSetID = 1)
```

## PubMed trend of enriched terms

Plots the number/proportion of publications trend based on the query result from PubMed Central.

```{r fig.width=10}
install.packages("europepmc", dependencies = T)
terms <- gse$Description[1:3]
pmcplot(terms, 2010:2018, proportion=FALSE)
```

```{r}
library(DOSE)
# Convert gene IDs for gseKEGG function
# We will lose some genes here because not all IDs will be converted
idsDvB<-bitr(names(original_gene_listDvB), fromType = "SYMBOL", toType = "ENTREZID", OrgDb=organism)

idsSexFvM<-bitr(names(original_gene_listSexFvM), fromType = "SYMBOL", toType = "ENTREZID", OrgDb=organism)
idsAgeOvY<-bitr(names(original_gene_listAgeOvY), fromType = "SYMBOL", toType = "ENTREZID", OrgDb=organism)
idsC1<-bitr(names(original_gene_listC1), fromType = "SYMBOL", toType = "ENTREZID", OrgDb=organism)
idsC2<-bitr(names(original_gene_listC2), fromType = "SYMBOL", toType = "ENTREZID", OrgDb=organism)
idsC3<-bitr(names(original_gene_listC3), fromType = "SYMBOL", toType = "ENTREZID", OrgDb=organism)

# remove duplicate IDS (here I use "ENSEMBL", but it should be whatever was selected as keyType)
dedup_idsDvB = idsDvB[!duplicated(idsDvB[c("ENTREZID")]),]

dedup_idsSexFvM = idsSexFvM[!duplicated(idsSexFvM[c("ENTREZID")]),]
dedup_idsAgeOvY = idsAgeOvY[!duplicated(idsAgeOvY[c("ENTREZID")]),]
dedup_idsC1 = idsC1[!duplicated(idsC1[c("ENTREZID")]),]
dedup_idsC2 = idsC2[!duplicated(idsC2[c("ENTREZID")]),]
dedup_idsC3 = idsC3[!duplicated(idsC3[c("ENTREZID")]),]

# Create a new dataframe df2 which has only the genes which were successfully mapped using the bitr function above
DvB_uniqgenes2 = DvB_uniqgenes[DvB_uniqgenes$external_gene_name %in% dedup_idsDvB$SYMBOL,]
DvB_uniqgenes2$SYMBOL = DvB_uniqgenes2$external_gene_name
DvB_genesEntrez <- left_join(DvB_uniqgenes2,dedup_idsDvB, by="SYMBOL")

SexFvM_uniqgenes2 = SexFvM_uniqgenes[SexFvM_uniqgenes$external_gene_name %in% dedup_idsSexFvM$SYMBOL,]
SexFvM_uniqgenes2$SYMBOL = SexFvM_uniqgenes2$external_gene_name
SexFvM_genesEntrez <- left_join(SexFvM_uniqgenes2,dedup_idsSexFvM, by="SYMBOL")

AgeOvY_uniqgenes2 = AgeOvY_uniqgenes[AgeOvY_uniqgenes$external_gene_name %in% dedup_idsAgeOvY$SYMBOL,]
AgeOvY_uniqgenes2$SYMBOL = AgeOvY_uniqgenes2$external_gene_name
AgeOvY_genesEntrez <- left_join(AgeOvY_uniqgenes2,dedup_idsAgeOvY, by="SYMBOL")

C1_uniqgenes2 = Cluster1_uniqgenes[Cluster1_uniqgenes$external_gene_name %in% dedup_idsC1$SYMBOL,]
C1_uniqgenes2$SYMBOL = C1_uniqgenes2$external_gene_name
C1_genesEntrez <- left_join(C1_uniqgenes2,dedup_idsC1, by="SYMBOL")

C2_uniqgenes2 = Cluster2_uniqgenes[Cluster2_uniqgenes$external_gene_name %in% dedup_idsC2$SYMBOL,]
C2_uniqgenes2$SYMBOL = C2_uniqgenes2$external_gene_name
C2_genesEntrez <- left_join(C2_uniqgenes2,dedup_idsC2, by="SYMBOL")

C3_uniqgenes2 = Cluster3_uniqgenes[Cluster3_uniqgenes$external_gene_name %in% dedup_idsC3$SYMBOL,]
C3_uniqgenes2$SYMBOL = C3_uniqgenes2$external_gene_name
C3_genesEntrez <- left_join(C3_uniqgenes2,dedup_idsC3, by="SYMBOL")

# Create a new column in df2 with the corresponding ENTREZ IDs
#DvB_uniqgenes2$Y = dedup_idsDvB$ENTREZID

# Create a vector of the gene universe
kegg_gene_listDvB <- DvB_genesEntrez$log2FoldChange

kegg_gene_listSexFvM <- SexFvM_genesEntrez$log2FoldChange
kegg_gene_listAgeOvY <- AgeOvY_genesEntrez$log2FoldChange
kegg_gene_listC1 <- C1_genesEntrez$log2FoldChange
kegg_gene_listC2 <- C2_genesEntrez$log2FoldChange
kegg_gene_listC3 <- C3_genesEntrez$log2FoldChange

# Name vector with ENTREZ ids
names(kegg_gene_listDvB) <- DvB_genesEntrez$ENTREZID

names(kegg_gene_listSexFvM) <- SexFvM_genesEntrez$ENTREZID
names(kegg_gene_listAgeOvY) <- AgeOvY_genesEntrez$ENTREZID
names(kegg_gene_listC1) <- C1_genesEntrez$ENTREZID
names(kegg_gene_listC2) <- C2_genesEntrez$ENTREZID
names(kegg_gene_listC3) <- C3_genesEntrez$ENTREZID

# omit any NA values 
kegg_gene_listDvB<-na.omit(kegg_gene_listDvB)

kegg_gene_listSexFvM<-na.omit(kegg_gene_listSexFvM)
kegg_gene_listAgeOvY<-na.omit(kegg_gene_listAgeOvY)
kegg_gene_listC1<-na.omit(kegg_gene_listC1)
kegg_gene_listC2<-na.omit(kegg_gene_listC2)
kegg_gene_listC3<-na.omit(kegg_gene_listC3)

# sort the list in decreasing order (required for clusterProfiler)
kegg_gene_listDvB = sort(kegg_gene_listDvB, decreasing = TRUE)

kegg_gene_listSexFvM = sort(kegg_gene_listSexFvM, decreasing = TRUE)
kegg_gene_listAgeOvY = sort(kegg_gene_listAgeOvY, decreasing = TRUE)
kegg_gene_listC1 = sort(kegg_gene_listC1, decreasing = TRUE)
kegg_gene_listC2 = sort(kegg_gene_listC2, decreasing = TRUE)
kegg_gene_listC3 = sort(kegg_gene_listC3, decreasing = TRUE)

## Create gseKEGG object
 
# **organism** KEGG Organism Code: The full list is here: https://www.genome.jp/kegg/catalog/org_list.html (need the 3 letter code). I define this as `kegg_organism` first, because it is used again below when making the pathview plots.  
# **nPerm** permutation numbers, the higher the number of permutations you set, the more accurate your results is, but it will also cost longer time for running permutation.  
# **minGSSize** minimal size of each geneSet for analyzing.   
# **maxGSSize** maximal size of genes annotated for testing.   
# **pvalueCutoff** pvalue Cutoff.   
# **pAdjustMethod** one of "holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr", "none".  
# **keyType** one of 'kegg', 'ncbi-geneid', 'ncib-proteinid' or 'uniprot'.  

kegg_organism = "hsa"
kk2DvB <- gseKEGG(geneList     = kegg_gene_listDvB,
               organism     = kegg_organism,
               nPerm        = 10000,
               minGSSize    = 3,
               maxGSSize    = 800,
               pvalueCutoff = 0.05,
               pAdjustMethod = "none",
               keyType       = "ncbi-geneid")

kk2SexFvM <- gseKEGG(geneList     = kegg_gene_listSexFvM,
               organism     = kegg_organism,
               nPerm        = 10000,
               minGSSize    = 3,
               maxGSSize    = 800,
               pvalueCutoff = 0.05,
               pAdjustMethod = "none",
               keyType       = "ncbi-geneid")
kk2AgeOvY <- gseKEGG(geneList     = kegg_gene_listAgeOvY,
               organism     = kegg_organism,
               nPerm        = 10000,
               minGSSize    = 3,
               maxGSSize    = 800,
               pvalueCutoff = 0.05,
               pAdjustMethod = "none",
               keyType       = "ncbi-geneid")
kk2C1 <- gseKEGG(geneList     = kegg_gene_listC1,
               organism     = kegg_organism,
               nPerm        = 10000,
               minGSSize    = 3,
               maxGSSize    = 800,
               pvalueCutoff = 0.05,
               pAdjustMethod = "none",
               keyType       = "ncbi-geneid")
kk2C2 <- gseKEGG(geneList     = kegg_gene_listC2,
               organism     = kegg_organism,
               nPerm        = 10000,
               minGSSize    = 3,
               maxGSSize    = 800,
               pvalueCutoff = 0.05,
               pAdjustMethod = "none",
               keyType       = "ncbi-geneid")
kk2C3 <- gseKEGG(geneList     = kegg_gene_listC3,
               organism     = kegg_organism,
               nPerm        = 10000,
               minGSSize    = 3,
               maxGSSize    = 800,
               pvalueCutoff = 0.05,
               pAdjustMethod = "none",
               keyType       = "ncbi-geneid")

save(kk2DvB, file =   "kk2DvB.RData")
save(kk2SexFvM, file ="kk2SexFvM.RData")
save(kk2AgeOvY, file ="kk2AgeOvY.RData")
save(kk2C1, file =    "kk2C1.RData")
save(kk2C2, file =    "kk2C2.RData")
save(kk2C3, file =    "kk2C3.RData")

load("kk2DvB.RData")
load("kk2SexFvM.RData")
load("kk2AgeOvY.RData")
load("kk2C1.RData")
load("kk2C2.RData")
load("kk2C3.RData")

png("k1.png", res = 300, width = 15, height = 15, units = "cm")

k1 <- dotplot(kk2DvB, showCategory = 10, title = "Enriched Pathways" , split=".sign") + facet_grid(.~.sign)+
  theme(axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))+
theme(legend.position="right", legend.text = element_text(size = 8),legend.title = element_text(size = 8)) + labs(title = "Enriched Pathways", y = "", x="") + theme(plot.title = element_text(size=10, hjust=0.5)) #+ scale_size_continuous(limits = c(0, 100), breaks = c(0, 25, 50, 75, 100))

k1

dev.off()


png("k2.png", res = 300, width = 15, height = 15, units = "cm")

k2 <- dotplot(kk2SexFvM, showCategory = 10, title = "Enriched Pathways" , split=".sign") + facet_grid(.~.sign)+
  theme(axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))+
theme(legend.position="right", legend.text = element_text(size = 8),legend.title = element_text(size = 8)) + labs(title = "Enriched Pathways", y = "", x="") + theme(plot.title = element_text(size=10, hjust=0.5)) #+ scale_size_continuous(limits = c(0, 100), breaks = c(0, 25, 50, 75, 100))

k2

dev.off()

png("k3.png", res = 300, width = 15, height = 15, units = "cm")

k3 <- dotplot(kk2AgeOvY, showCategory = 10, title = "Enriched Pathways" , split=".sign") + facet_grid(.~.sign)+
  theme(axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))+
theme(legend.position="right", legend.text = element_text(size = 8),legend.title = element_text(size = 8)) + labs(title = "Enriched Pathways", y = "", x="") + theme(plot.title = element_text(size=10, hjust=0.5)) #+ scale_size_continuous(limits = c(0, 100), breaks = c(0, 25, 50, 75, 100))

k3

dev.off()

png("k4.png", res = 300, width = 15, height = 15, units = "cm")

k4 <- dotplot(kk2C1, showCategory = 10, title = "Enriched Pathways" , split=".sign") + facet_grid(.~.sign)+
  theme(axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))+
theme(legend.position="right", legend.text = element_text(size = 8),legend.title = element_text(size = 8)) + labs(title = "Enriched Pathways", y = "", x="") + theme(plot.title = element_text(size=10, hjust=0.5)) #+ scale_size_continuous(limits = c(0, 100), breaks = c(0, 25, 50, 75, 100))

k4

dev.off()

png("k5.png", res = 300, width = 15, height = 15, units = "cm")

k5 <-dotplot(kk2C2, showCategory = 10, title = "Enriched Pathways" , split=".sign") + facet_grid(.~.sign)+
  theme(axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))+
theme(legend.position="right", legend.text = element_text(size = 8),legend.title = element_text(size = 8)) + labs(title = "Enriched Pathways", y = "", x="") + theme(plot.title = element_text(size=10, hjust=0.5)) #+ scale_size_continuous(limits = c(0, 100), breaks = c(0, 25, 50, 75, 100))

k5

dev.off()

png("k6.png", res = 300, width = 15, height = 15, units = "cm")

k6 <- dotplot(kk2C3, showCategory = 10, title = "Enriched Pathways" , split=".sign") + facet_grid(.~.sign)+
  theme(axis.text.x=element_text(size=7),axis.text.y=element_text(size=7))+
theme(legend.position="right", legend.text = element_text(size = 8),legend.title = element_text(size = 8)) + labs(title = "Enriched Pathways", y = "", x="") + theme(plot.title = element_text(size=10, hjust=0.5)) #+ scale_size_continuous(limits = c(0, 100), breaks = c(0, 25, 50, 75, 100))

k6

dev.off()

```

# NOTA: Modificar esto para que corra en una sola figura. 

Problemas con el volcano plot y los dotplot

```{r}

png("figure3.png", res = 300, width = 50, height = 50, units = "cm")

plot_grid(DvB1, dp1, cnet1, k1,
          ncol = 2,
          labels = c("A","B", "C", "D"),
          label_size = 5,
          label_fontfamily = "serif"
          )

dev.off()


png("patchFig3.png", res = 300, width = 100, height = 100, units = "cm")

(DvB1 | dp1) /
(cnet1 | k1)

dev.off()



png("figure4.png", res = 300, width = 50, height = 50, units = "cm")

plot_grid(SexFvM, dp2, cnet2, k2,
          ncol = 2,
          labels = c("A","B", "C", "D"),
          label_size = 5,
          label_fontfamily = "serif"
          )

dev.off()

png("figure5.png", res = 300, width = 30, height = 50, units = "cm")

plot_grid(AgeOvY, dp3, cnet3, k3,
          ncol = 2,
          labels = c("A","B", "C", "D"),
          label_size = 5,
          label_fontfamily = "serif"
          )

dev.off()


png("figure6.png", res = 300, width = 30, height = 50, units = "cm")

plot_grid(dp4, k4, dp5, k5, dp6, k6,
          ncol = 2,
          labels = c("A","B", "C", "D", "E", "F"),
          label_size = 5,
          label_fontfamily = "serif"
          )

dev.off()
```

